### PR TITLE
Add return-from-error issue

### PR DESCRIPTION
**Pull request by yitzchak**

### CODE REVIEW

File: wscl-issues/proposed/return-from-error [hunk 1]
```diff
@@ -0,0 +1,115 @@
+Issue:          RETURN-FROM-ERROR
+Forum:          Cleanup
+Category:       CLARIFICATION
+Status:         proposed
+Edit History:   12-Sep-21, Version 1 by Tarn W. Burton.
+References:     RETURN-FROM
```
Comment (scymtym) [inline wscl-issues/proposed/return-from-error:6]: Should probably reference `program-error` since that entry specifically mentions "… result from naming a go tag or a block tag that is not lexically apparent …"

File: wscl-issues/proposed/return-from-error [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  special operator RETURN-FROM indicates no exceptional situations.
+
+Proposal (RETURN-FROM-ERROR:SIGNAL-ERROR-IN-SAFE-CODE):
+
+  This proposal changes the description of the RETURN-FROM special
+  operator, so that the section "Exceptional Situations" contains the
+  following text:
+
+  "The special operator RETURN-FROM should signal an error of type
+  TYPE-ERROR if the NAME argument is not a symbol. A PROGRAM-ERROR
```
Comment (scymtym) [inline wscl-issues/proposed/return-from-error:20]: Do we have a precedent or a policy for type error vs program error in this situation? When a non-evaluated sub-expression of a special form is of the wrong type, that is arguably both a type error and a program error.

**Review by scymtym — CHANGES_REQUESTED** [review]
Suggestions and a question in commit comments.

File: wscl-issues/proposed/return-from-error [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  should be signaled if the NAME argument refers to a block tag that
+  is not lexically apparent."
```
[the remark this answers is not in the document]
Comment (yitzchak) [reply wscl-issues/proposed/return-from-error:22]: Sounds good.

File: wscl-issues/proposed/return-from-error (hunk 1 above)
Comment (yitzchak) [reply wscl-issues/proposed/return-from-error:6]: Sounds good

File: wscl-issues/proposed/return-from-error (hunk 2 above)
Comment (yitzchak) [reply wscl-issues/proposed/return-from-error:20]: I'll look to see there is one.

My thinking was that validating the types of the arguments would initially be done with `check-type` (obviously not optimized). Once arguments are known to be valid then a symbol that is not an actual block name is  a program error.

Comment (Bike) [reply wscl-issues/proposed/return-from-error]: i flipped through the special operators and none of them are required to signal anything for this kind of syntax error that i can see. maybe some macros do, i don't know.
me, i'm not sure how useful it would be to have it specified more closely than `ERROR`. any compiler worth its salt is going to at least full-warn on code like this, and it's hard to imagine any sensible handling of/recovery from the error by users.

Comment (yitzchak) [reply wscl-issues/proposed/return-from-error]: I can see how there is not much of a useful difference between the `type-error` and `program-error`, but what about that fact that the docs for `program-error` say that it is going to be signaled when there is no lexically available block by a specific name? Wouldn't saying that `error` is signaled contradict that? 

As an alternative one could rewrite the exception section to say "The special  operator RETURN-FROM should signal an error of type PROGRAM-ERROR  if no BLOCK form named NAME lexically encloses the occurrence of  the RETURN-FROM form.  The special operator RETURN-FROM may signal an error of type  TYPE-ERROR if the NAME argument is not a symbol. "

Basically replace the "should" in the type-error sentence with "may" so that implementations that just signal `program-error` and those that chose to type check the name are both conformant.

Just a thought.

Comment (Bike) [reply wscl-issues/proposed/return-from-error]: I mean, if I had to pick one or the other I'd go with at least making it a `program-error`. Program errors are related to incorrect syntax, so I'd expect something like `(let (a 4) a)` to be a program error, even though this could alternately be described as a type error ("`4` is not of type `(or symbol (cons symbol))`" for example).

Technically speaking I don't think that specifying only an error of type `error` would contradict something else saying it's a `program-error`, since any program error is also of type error, but it would be meaninglessly confusing, yeah.

Comment (yitzchak) [reply wscl-issues/proposed/return-from-error]: That makes sense. I was going to write similar proposals for `block`, etc. Just to clarify, do you think that `(block "fu")` is more of a `program-error` vs a `type-error`?

Comment (Bike) [reply wscl-issues/proposed/return-from-error]: yeah. i mean, it could be both, but i think program error is more important. a program error has the specific meaning of "the program text is wrong somehow", whereas a type error could mean any number of things.

Comment (yitzchak) [reply wscl-issues/proposed/return-from-error]: Seems like we have three different possibilities for the updated text:

1. The special operator RETURN-FROM should signal an error of type   TYPE-ERROR if the NAME argument is not a symbol. The special   operator RETURN-FROM should signal an error of type PROGRAM-ERROR   if no BLOCK form named NAME lexically encloses the occurrence of   the RETURN-FROM form.
2. The special operator RETURN-FROM will signal an error of type PROGRAM-ERROR if no BLOCK form named NAME lexically encloses the occurrence of the RETURN-FROM form. The special operator RETURN-FROM may signal an error of type TYPE-ERROR if the NAME argument is not a symbol.
3. The special operator RETURN-FROM will signal an error of type PROGRAM-ERROR if no BLOCK form named NAME lexically encloses the occurrence of the RETURN-FROM form or if NAME is not a symbol.

Under option 1 none of the current implementations are conformant. Under option 2 and 3, CCL, CLASP, CMU, LWPE and SBCL are conformant. Under option 3, ABCL and CLISP only have to change the error signaled for `(one a b)` to be conformant. For this option they do not need change the error signalled for `(one "a" a)` in other to conformant.

@scymtym @Bike Thoughts?

Comment (scymtym) [reply wscl-issues/proposed/return-from-error]: My head-canon has been:
* Wrong character syntax => `syntax-error` from `read`
* Wrong s-expression syntax in special operator => `compile` indicates failure, `program-error` at runtime, arranged by `compile`
* Wrong s-expression syntax in macro => some non-recoverable error at compile-time
* Everything else => error at runtime, including `control-error` for invalid control flow that is not apparent from the lexical program structure

So my preference is 3.

Spelling this out, I now think the "dynamic" exceptional situation should also be mentioned. The `control-error` entry says:

>  The errors that result from giving throw a tag that is not active or from giving go or return-from a tag that is no longer dynamically available are of type control-error.

Comment (scymtym) [reply wscl-issues/proposed/return-from-error]: Thinking about "if NAME is not a symbol" more, I'm not sure anymore that it makes sense to include this as an exceptional situation. The required type of the name is specified in the "Arguments and Values" section. The [description of the "Arguments and Values" section](http://www.lispworks.com/documentation/HyperSpec/Body/01_ddc.htm) states
> Except as explicitly specified otherwise, the consequences are undefined if these type restrictions are violated.

I think it would make more sense to reserve "Exceptional Situations" for situations that arise at runtime or in a more complex way than a violation of the syntax of the special operator ("more complex way" could be something like invalid lexical surroundings or something else not confined to the form in question). It could make sense to specify in the description of the "Arguments and Values" section what should happen if a special form does not conform to the syntax of "its" special operator.

Comment (yitzchak) [reply wscl-issues/proposed/return-from-error]: Well, that would simplify things considerably. I guess I will remove the test `(one a "b")` since that is really a test of syntax and then leave that to a proposal for "Arguments and Values" section?

Comment (scymtym) [reply wscl-issues/proposed/return-from-error]: I would like to hear @Bike's thoughts.

Comment (Bike) [reply wscl-issues/proposed/return-from-error]: We already have several issues trying to add "if one of the argument types is wrong, a type error should be signaled" type lines to descriptions of functions. If that makes sense in the "Exceptional Situations" I think it makes sense here as well.

But also, I'm honestly just not very concerned about what needs to happen if return-from gets a non-symbol. I think it should signal an error, and probably a program-error, but like I said it's hard to imagine any response other than the programmer going "oh, whoops, typo" and fixing it. The more I talk about it the more I feel like I'm bikeshedding. But of the three options I would also pick 3.

I think the lexical problem, i.e. when the name is properly a symbol but there's no block lexically visible, is more important, and the proposal to do a program error there is fine and I think we all agree on that.

The dynamic problem scymtym brings up is even more important, in my view, since that can actually happen in a non-obvious way and warrants giving the programmer as much information as possible to try to figure out what happened. But I think that's sort of a separate issue given that it's obviously not a syntax error, and probably brings in some other issues (e.g.: is an implementation permitted to leave some exit points valid that are not required to be? The standard makes the consequences undefined, so an implementation can allow the "unstoppable loops" mentioned in the exit-extent:medium issue.) so this might not be the place.

scymtym also mentions the condition behavior of the compiler. That's something I've thought about a bit - like whether there could be a defined condition hierarchy for the compiler to indicate problems like this - which could be very useful for editors and stuff. But it's probably out of scope for this issue and even WSCL entirely.

Comment (scymtym) [reply wscl-issues/proposed/return-from-error]: > We already have several issues trying to add "if one of the argument types is wrong, a type error should be signaled" type lines to descriptions of functions. If that makes sense in the "Exceptional Situations" I think it makes sense here as well.

I'm not sure whether it makes sense to handle functions and special operators in the same way. For functions, argument type errors are never also (s-expression) syntax errors since function arguments are always evaluated and the type mismatch is thus between the expected type and a runtime value, not a part of the program.

I'm leaning towards not discussing "if NAME is not a symbol" in "Exceptional Situations" because
* The type mismatch does not happen at runtime the way it works for function arguments
* If we start discussing the handling of type constraints imposed on *unevaluated* arguments by the "Arguments and Values" section, where should we stop? Using something from an earlier comment, should we then describe the runtime behavior of `(let 1 …)`, `(let (1) …)`, etc in the "Exceptional Situations" of `let`?
  * One problem would be the explosion of exceptional situations
  * Another problem is possibly forcing compilers to recover and complete the compilation process to the point where the mandated error can be signaled at runtime

I *do* think that a `program-error` is the right error to be signaled at runtime if compilation finishes, but I would prefer to handle this kind of syntax errors (again, I'm specifically referring to unevaluated arguments of special operators) in some other way.

> But also, I'm honestly just not very concerned about what needs to happen if return-from gets a non-symbol. I think it should signal an error, and probably a program-error, but like I said it's hard to imagine any response other than the programmer going "oh, whoops, typo" and fixing it. The more I talk about it the more I feel like I'm bikeshedding. But of the three options I would also pick 3.

I think the discussion is important because the we will probably apply the conclusion to all special operators.

> I think the lexical problem, i.e. when the name is properly a symbol but there's no block lexically visible, is more important, and the proposal to do a program error there is fine and I think we all agree on that.

Yes, I'm definitely in favor of that.

I agree that the dynamic aspect and a wider consideration of the compiler's behavior are out of scope for this issue.

Comment (yitzchak) [reply wscl-issues/proposed/return-from-error]: Updated for option 3 without mentioning the symbol syntax error.

**Review by scymtym — APPROVED** [review]

**Comment by scymtym** [comment]
Looks good to me.
